crates/modelardb_compression/src/types.rs

```diff
@@ -0,0 +1,622 @@
+    pub fn try_new(percentage: f32) -> Result<Self, ModelarDbError> {
+        if !(0.0..=100.0).contains(&percentage) {
+            Err(ModelarDbError::CompressionError(
+                "Error bound must be a value from 0.0% to 100.0%.".to_owned(),
+            ))
+        } else {
+            Ok(Self(percentage))
+        }
+    }
+
+    /// Consumes `self`, returning the error bound as a [`f32`].
+    pub fn into_inner(self) -> f32 {
+        self.0
+    }
+}
+
+/// Enable equal and not equal for [`ErrorBound`] and [`f32`].
+impl PartialEq<ErrorBound> for f32 {
+    fn eq(&self, other: &ErrorBound) -> bool {
+        self.eq(&other.0)
+    }
+}
+
+/// Enable less than and greater than for [`ErrorBound`] and [`f32`].
+impl PartialOrd<ErrorBound> for f32 {
+    fn partial_cmp(&self, other: &ErrorBound) -> Option<Ordering> {
+        self.partial_cmp(&other.0)
+    }
+}
+
+/// A model being built from an uncompressed segment using the potentially lossy model types in
+/// [`models`]. Each of the potentially lossy model types is used to fit models to the data points,
+/// and then the model that uses the fewest number of bytes per value is selected.
+pub(crate) struct ModelBuilder {
+    /// Index of the first data point in `uncompressed_timestamps` and `uncompressed_values` the
+    /// compressed model represents values for.
+    start_index: usize,
+    /// Constant function that currently represents the values in `uncompressed_values` from
+    /// `start_index` to `start_index` + `pmc_mean.length`.
+    pmc_mean: PMCMean,
+    /// Indicates if `pmc_mean` could represent all values in `uncompressed_values` from
+    /// `start_index` to `current_index` in `new()`.
+    pmc_mean_could_fit_all: bool,
+    /// Linear function that represents the values in `uncompressed_values` from `start_index` to
+    /// `start_index` + `swing.length`.
+    swing: Swing,
+    /// Indicates if `swing` could represent all values in `uncompressed_values` from `start_index`
+    /// to `current_index` in `new()`.
+    swing_could_fit_all: bool,
+}
+
+impl ModelBuilder {
+    /// Create a model that represents a sub-sequence of uncompressed values that starts at
+    /// `start_index`, is within `error_bound`, and uses the fewest number of bytes per value.
+    pub(crate) fn new(start_index: usize, error_bound: ErrorBound) -> Self {
+        Self {
+            start_index,
+            pmc_mean: PMCMean::new(error_bound),
+            pmc_mean_could_fit_all: true,
+            swing: Swing::new(error_bound),
+            swing_could_fit_all: true,
+        }
+    }
+
+    /// Attempt to update the current models to also represent the `value` of
+    /// the data point collected at `timestamp`.
+    pub(crate) fn try_to_update_models(&mut self, timestamp: Timestamp, value: Value) {
+        debug_assert!(
+            self.can_fit_more(),
+            "The current models cannot be fitted to additional data points."
+        );
```

Comment:
This is good idea, but I made `try_to_update_models()` return `bool` instead of `Result` as we use the value it returns to drive a loop and thus have no need for a value in `Ok` or `Err`. So I ended up emulating a boolean by returning `Result<(), ()>` and then have the loop in `compression.rs` call `is_ok()`. Surprisingly for me the code generated for a function returning `bool` and `Result::<(), ()>` is the [same with all optimizations enabled](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:18,endLineNumber:25,positionColumn:18,positionLineNumber:25,selectionStartColumn:18,selectionStartLineNumber:25,startColumn:18,startLineNumber:25),source:'use+std::env%3B%0A%0Apub+fn+can_divide_by_two_bool(i:+u64)+-%3E+bool+%7B%0A++++if+i+%25+37+%3D%3D+0+%7B%0A++++++++true%0A++++%7D+else+%7B%0A++++++++false%0A++++%7D%0A%7D%0A%0Apub+fn+can_divide_by_two_result(i:+u64)+-%3E+Result%3C(),+()%3E+%7B%0A++++if+i+%25+37+%3D%3D+0+%7B%0A++++++++Ok(())%0A++++%7D+else+%7B%0A++++++++Err(())%0A++++%7D%0A%7D%0A%0A%0Apub+fn+main()+-%3E+usize+%7B%0A++++let+string+%3D+env::args().next().unwrap()%3B%0A++++let+number+%3D+string.parse::%3Cu64%3E().unwrap()%3B+%0A%0A++++if+can_divide_by_two_bool(number)+%7B%0A++++++++return+1%3B%0A++++%7D%0A%0A++++if+can_divide_by_two_result(number).is_ok()+%7B%0A++++++++return+2%3B%0A++++%7D%0A%0A++++return+3%3B%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1680,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+opt-level%3D3',overrides:!(),selection:(endColumn:15,endLineNumber:71,positionColumn:15,positionLineNumber:71,selectionStartColumn:15,selectionStartLineNumber:71,startColumn:15,startLineNumber:71),source:1),l:'5',n:'0',o:'+rustc+1.68.0+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4). Although I do not know if `is_ok()` can always be optimized completely away, so it may have some runtime overhead compared to a boolean.